### PR TITLE
[ADD] Automatically enabled indices for foreign keys

### PR DESCRIPTION
**Pull request by StefanRijnhart**
Description of the issue/feature this PR addresses: we got fed up with finding and fixing bottlenecks in database performance by adding databases on an ad hoc basis.

Current behavior before PR: Indexes on Many2one fields were only created when index = True in the code.

Desired behavior after PR is merged: indexes on Many2one fields are created unless index = False in the code.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by StefanRijnhart** [comment]
cc @lembregtse

**Comment by feketemihai** [comment]
@StefanRijnhart super...this will improve for sure performance.. Thanks.

**Comment by mart-e** [comment]
Hi,

I am not sure setting an index on every many2one will improve the performances.
Adding an index will make the lookup faster but will slow down the insert or update (when the index is recomputed). Adding an index on a field you never search on will be counterproductive.

**Comment by rousseldenis** [comment]
@mart-e I think you should consider this as, from my experience, adding indexes like that never drop down performances or don't affect database such a manner writes and creates are hugely degraded.

If you place yourself at the user side, you expect searches the fastest. Because user does not want to wait... You improve user xp with this.

And if you put indexes on things you don't search, I bet you won't have so much data, so the impact is neglectable.

Maybe this is too intrusive. @rco-odoo maybe another elegant solution for this?

**Comment by wtaferner** [comment]
@StefanRijnhart @mart-e 
Actually I think it would be best practice to somehow monitor the type of usage in a module (maybe even on framework side) and based on this give recommendations or even trigger automatically to add/remove an index during installation/upgrade relying on this data.
I guess PostgreSQL possibly has all the statistics which is needed even without analyzing in Odoo itself what is done and how often, etc., but not sure...at least if it is monitored in Odoo we have the application knowing it's type of usage best giving recommendation or even automatically adapting indices on install/upgrade. No need for inheriting or customizing for each instance, model and field to change the database.

What do you think? Sounds neat...just don't know if it is as easy as it sounds :-)

@rco-odoo 
Any comments?

**Comment by odony** [comment]
Indeed, I don't think it would be a good idea or an interesting trade-off:

- Clearly not all FK are relevant for search queries, and the ones that are can be quite different on a case-by-case basis, for each company / deployment
- Indices contain pure duplicate data, and significantly contribute to database size, making maintenance vacuuming slower and more expensive, backups harder and slower to take, etc. When they are unused, this is pure waste.
- Maintaining indices will significantly slow down INSERT, UPDATE and DELETE queries, and cause more data bloat. Are we ready to take a global performance hit to requests that alter data? (@StefanRijnhart's sample test gave a 4% time hit on a near empty database, but includes index creation, so may not be very relevant for real-world use)
- Simple indices are often insufficient or inappropriate to solve performance problems on large volumes. Composite and partial indices are often required (and not just on FKs), and require manual analysis.
- The only operation that will always benefit from indexed FKs is DELETE (which requires inverse FK lookups), but this is the least common type of transaction. The effect is amplified for master data which is typically heavily cross-referenced by FKs, but is also the least likely to be hard-deleted.

To put this into perspective, here are some relevant numbers [1]:
- Number of columns on a 11.0 runbot DB (-all): ~ 10000
- Number of FKs: ~ 3400 (34% of all columns seem to be Many2Ones)
- Number of current indices on FKs: ~ 600 (18% of FKs)
- Size with current indices: Total bytes: 164 MB, Table bytes: 91 MB, Index bytes: 73 MB (45% of total DB size)
- Sizes of Odoo.com DB (including a few extra partial/composite indices): Total bytes: 183 GB, Table bytes: 76 GB, Index bytes: 107 GB (nearly 60% of total DB size)

In a nutshell, it looks like this change would multiply by 5 the number of FK indexes, create a significant overhead for data-altering queries and increase database sizes (duplicate data for 1/3 of all columns), all of it without clearly addressing a specific performance issue.

I'm much more in favor of adding indices on a case-by-case basis when we have a clear and generic rationale, and leaving custom indices and usage-specific optimizations to admins.

[1] here are [the queries I used](https://gist.github.com/odony/11e146371e25d74561e15c0fa399c04d), in case I got the estimates wrong

**Comment by StefanRijnhart** [comment]
Thanks for your insights everyone, especially @odony. I understand this is not for everyone and I will carefully examine the effects on my own projects. Closing this.
